lib/Core/Executor.cpp

```diff
@@ -905,14 +905,21 @@ Executor::fork(ExecutionState &current, ref<Expr> condition, bool isInternal) {
       }
     }
 
+    // We don't add a constraint explicitly yet
```

Comment:
@MartinNowack @ccadar This is interesting indeed. :) Let me just make sure my understanding is correct. Essentially, there are 3 paths/traces in this program: an infinite one, a failing one, and a successful one. Based on the current strategy, KLEE (with the first patch) chooses to explore the infinite one first. As it's doing this, it hits several symbolic branches, calls the solver to check feasibility each time, but there is always just a single feasible branch. As a consequence, it never "re-schedules" the other branch that would let it explore the two other feasible paths.

If this is indeed what's happening then I can understand both of your points. This seems to boil down to a somewhat philosophical question about when the search strategy should get to interfere. Martin's second solution considers each static branch in the code as a good point, while Cristian is arguing for only consulting the search strategy at *dynamic* branches (i.e., where both sides are feasible). BTW, is it correct that this example would also be problematic with a random strategy since it may also choose the infinite path initially and get stuck?

I would argue that from a user perspective it would be desirable if KLEE would be able to find the assertion violation. It seems like Cristian's suggestion about consulting the strategy on regular intervals would be a very pragmatic solution to achieve this. In fact, it would make KLEE much more resilient to non-terminating traces in general (see Cristian's example about a "concrete infinite loop").

Would it make sense to, at least optionally, add such a feature?